// @solana-mobile/mobile-wallet-adapter-protocol/lib/cjs/index.native.js
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
let react_native = require("react-native");
let _solana_codecs_strings = require("@solana/codecs-strings");
let js_base64 = require("js-base64");
let _solana_wallet_standard_util = require("@solana/wallet-standard-util");
//#region src/errors.ts
const SolanaMobileWalletAdapterErrorCode = {
	ERROR_ASSOCIATION_PORT_OUT_OF_RANGE: "ERROR_ASSOCIATION_PORT_OUT_OF_RANGE",
	ERROR_REFLECTOR_ID_OUT_OF_RANGE: "ERROR_REFLECTOR_ID_OUT_OF_RANGE",
	ERROR_FORBIDDEN_WALLET_BASE_URL: "ERROR_FORBIDDEN_WALLET_BASE_URL",
	ERROR_SECURE_CONTEXT_REQUIRED: "ERROR_SECURE_CONTEXT_REQUIRED",
	ERROR_SESSION_CLOSED: "ERROR_SESSION_CLOSED",
	ERROR_SESSION_TIMEOUT: "ERROR_SESSION_TIMEOUT",
	ERROR_WALLET_NOT_FOUND: "ERROR_WALLET_NOT_FOUND",
	ERROR_INVALID_PROTOCOL_VERSION: "ERROR_INVALID_PROTOCOL_VERSION",
	ERROR_BROWSER_NOT_SUPPORTED: "ERROR_BROWSER_NOT_SUPPORTED",
	ERROR_LOOPBACK_ACCESS_BLOCKED: "ERROR_LOOPBACK_ACCESS_BLOCKED",
	ERROR_ASSOCIATION_CANCELLED: "ERROR_ASSOCIATION_CANCELLED"
};
var SolanaMobileWalletAdapterError = class extends Error {
	data;
	code;
	constructor(...args) {
		const [code, message, data] = args;
		super(message);
		this.code = code;
		this.data = data;
		this.name = "SolanaMobileWalletAdapterError";
	}
};
const SolanaMobileWalletAdapterProtocolErrorCode = {
	ERROR_AUTHORIZATION_FAILED: -1,
	ERROR_INVALID_PAYLOADS: -2,
	ERROR_NOT_SIGNED: -3,
	ERROR_NOT_SUBMITTED: -4,
	ERROR_TOO_MANY_PAYLOADS: -5,
	ERROR_ATTEST_ORIGIN_ANDROID: -100
};
var SolanaMobileWalletAdapterProtocolError = class extends Error {
	data;
	code;
	jsonRpcMessageId;
	constructor(...args) {
		const [jsonRpcMessageId, code, message, data] = args;
		super(message);
		this.code = code;
		this.data = data;
		this.jsonRpcMessageId = jsonRpcMessageId;
		this.name = "SolanaMobileWalletAdapterProtocolError";
	}
};
//#endregion
//#region src/codegenSpec/NativeSolanaMobileWalletAdapter.ts
var NativeSolanaMobileWalletAdapter_default = react_native.TurboModuleRegistry.getEnforcing("SolanaMobileWalletAdapter");
//#endregion
//#region src/base58Utils.ts
function fromUint8Array(byteArray) {
	return (0, _solana_codecs_strings.getBase58Decoder)().decode(byteArray);
}
function base64ToBase58(base64EncodedString) {
	return fromUint8Array((0, js_base64.toUint8Array)(base64EncodedString));
}
//#endregion
//#region src/createSIWSMessage.ts
function createSIWSMessage(payload) {
	return (0, _solana_wallet_standard_util.createSignInMessageText)(payload);
}
function createSIWSMessageBase64Url(payload) {
	return (0, js_base64.encode)(createSIWSMessage(payload)).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}
//#endregion
//#region src/types.ts
const SolanaSignTransactions = "solana:signTransactions";
const SolanaCloneAuthorization = "solana:cloneAuthorization";
const SolanaSignInWithSolana = "solana:signInWithSolana";
//#endregion
//#region src/createMobileWalletProxy.ts
/**
* Creates a {@link MobileWallet} proxy that handles backwards compatibility and API to RPC conversion.
*
* @param protocolVersion the protocol version in use for this session/request
* @param protocolRequestHandler callback function that handles sending the RPC request to the wallet endpoint.
* @returns a {@link MobileWallet} proxy
*/
function createMobileWalletProxy(protocolVersion, protocolRequestHandler) {
	return new Proxy({}, {
		get(target, p) {
			if (p === "then") return null;
			if (target[p] == null) target[p] = async function(inputParams) {
				const { method, params } = handleMobileWalletRequest(p, inputParams, protocolVersion);
				const result = await protocolRequestHandler(method, params);
				if (method === "authorize" && params.sign_in_payload && !result.sign_in_result) result.sign_in_result = await signInFallback(params.sign_in_payload, result, protocolRequestHandler);
				return handleMobileWalletResponse(p, result, protocolVersion);
			};
			return target[p];
		},
		defineProperty() {
			return false;
		},
		deleteProperty() {
			return false;
		}
	});
}
/**
* Handles all {@link MobileWallet} API requests and determines the correct MWA RPC method and params to call.
* This handles backwards compatibility, based on the provided @protocolVersion.
*
* @param methodName the name of {@link MobileWallet} method that was called
* @param methodParams the parameters that were passed to the method
* @param protocolVersion the protocol version in use for this session/request
* @returns the RPC request method and params that should be sent to the wallet endpoint
*/
function handleMobileWalletRequest(methodName, methodParams, protocolVersion) {
	let params = methodParams;
	let method = methodName.toString().replace(/[A-Z]/g, (letter) => `_${letter.toLowerCase()}`).toLowerCase();
	switch (methodName) {
		case "authorize": {
			const authorizeParams = params;
			let { chain } = authorizeParams;
			if (protocolVersion === "legacy") {
				switch (chain) {
					case "solana:testnet":
						chain = "testnet";
						break;
					case "solana:devnet":
						chain = "devnet";
						break;
					case "solana:mainnet":
						chain = "mainnet-beta";
						break;
					default: chain = authorizeParams.cluster;
				}
				authorizeParams.cluster = chain;
				params = authorizeParams;
			} else {
				switch (chain) {
					case "testnet":
					case "devnet":
						chain = `solana:${chain}`;
						break;
					case "mainnet-beta":
						chain = "solana:mainnet";
						break;
				}
				authorizeParams.chain = chain;
				params = authorizeParams;
			}
		}
		case "reauthorize": {
			const { auth_token, identity } = params;
			if (auth_token) switch (protocolVersion) {
				case "legacy":
					method = "reauthorize";
					params = {
						auth_token,
						identity
					};
					break;
				default:
					method = "authorize";
					break;
			}
			break;
		}
	}
	return {
		method,
		params
	};
}
/**
* Handles all {@link MobileWallet} API responses and modifies the response for backwards compatibility, if needed
*
* @param method the {@link MobileWallet} method that was called
* @param response the original response that was returned by the method call
* @param protocolVersion the protocol version in use for this session/request
* @returns the possibly modified response
*/
function handleMobileWalletResponse(method, response, protocolVersion) {
	switch (method) {
		case "getCapabilities": {
			const capabilities = response;
			switch (protocolVersion) {
				case "legacy": {
					const features = [SolanaSignTransactions];
					if (capabilities.supports_clone_authorization === true) features.push(SolanaCloneAuthorization);
					return {
						...capabilities,
						features
					};
				}
				case "v1": return {
					...capabilities,
					supports_sign_and_send_transactions: true,
					supports_clone_authorization: capabilities.features.includes(SolanaCloneAuthorization)
				};
			}
		}
	}
	return response;
}
async function signInFallback(signInPayload, authorizationResult, protocolRequestHandler) {
	const domain = signInPayload.domain ?? window.location.host;
	const address = authorizationResult.accounts[0].address;
	const siwsMessage = createSIWSMessageBase64Url({
		...signInPayload,
		domain,
		address: base64ToBase58(address)
	});
	const signedPayload = (0, js_base64.toUint8Array)((await protocolRequestHandler("sign_messages", {
		addresses: [address],
		payloads: [siwsMessage]
	})).signed_payloads[0]);
	const signedMessage = (0, js_base64.fromUint8Array)(signedPayload.slice(0, signedPayload.length - 64));
	const signature = (0, js_base64.fromUint8Array)(signedPayload.slice(signedPayload.length - 64));
	return {
		address,
		signed_message: signedMessage.length == 0 ? siwsMessage : signedMessage,
		signature
	};
}
//#endregion
//#region src/__forks__/react-native/transact.ts
react_native.AppRegistry.registerHeadlessTask("SolanaMobileWalletAdapterSessionBackgroundTask", () => {
	return async () => {};
});
const LINKING_ERROR = "The package 'solana-mobile-wallet-adapter-protocol' doesn't seem to be linked. Make sure: \n\n- You rebuilt the app after installing the package\n- If you are using Lerna workspaces\n  - You have added `@solana-mobile/mobile-wallet-adapter-protocol` as an explicit dependency, and\n  - You have added `@solana-mobile/mobile-wallet-adapter-protocol` to the `nohoist` section of your package.json\n- You are not using Expo managed workflow\n";
const SolanaMobileWalletAdapter = react_native.Platform.OS === "android" && NativeSolanaMobileWalletAdapter_default ? NativeSolanaMobileWalletAdapter_default : new Proxy({}, { get() {
	throw new Error(react_native.Platform.OS !== "android" ? "The package `solana-mobile-wallet-adapter-protocol` is only compatible with React Native Android" : LINKING_ERROR);
} });
function getErrorMessage(e) {
	switch (e.code) {
		case "ERROR_WALLET_NOT_FOUND": return "Found no installed wallet that supports the mobile wallet protocol.";
		default: return e.message;
	}
}
function handleError(e) {
	if (e instanceof Error) {
		const reactNativeError = e;
		switch (reactNativeError.code) {
			case void 0: throw e;
			case "JSON_RPC_ERROR": {
				const details = reactNativeError.userInfo;
				throw new SolanaMobileWalletAdapterProtocolError(0, details.jsonRpcErrorCode, e.message);
			}
			default: throw new SolanaMobileWalletAdapterError(reactNativeError.code, getErrorMessage(reactNativeError), reactNativeError.userInfo);
		}
	}
	throw e;
}
async function transact(callback, config) {
	let didSuccessfullyConnect = false;
	try {
		const sessionProperties = await SolanaMobileWalletAdapter.startSession(config);
		didSuccessfullyConnect = true;
		return await callback(createMobileWalletProxy(sessionProperties.protocol_version, async (method, params) => {
			try {
				return SolanaMobileWalletAdapter.invoke(method, params);
			} catch (e) {
				return handleError(e);
			}
		}));
	} catch (e) {
		return handleError(e);
	} finally {
		if (didSuccessfullyConnect) await SolanaMobileWalletAdapter.endSession();
	}
}
//#endregion
exports.SolanaCloneAuthorization = SolanaCloneAuthorization;
exports.SolanaMobileWalletAdapterError = SolanaMobileWalletAdapterError;
exports.SolanaMobileWalletAdapterErrorCode = SolanaMobileWalletAdapterErrorCode;
exports.SolanaMobileWalletAdapterProtocolError = SolanaMobileWalletAdapterProtocolError;
exports.SolanaMobileWalletAdapterProtocolErrorCode = SolanaMobileWalletAdapterProtocolErrorCode;
exports.SolanaSignInWithSolana = SolanaSignInWithSolana;
exports.SolanaSignTransactions = SolanaSignTransactions;
exports.transact = transact;
