// @solana-mobile/mobile-wallet-adapter-protocol-web3js/lib/cjs/index.native.js
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
//#region \0rolldown/runtime.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));
//#endregion
let _solana_web3_js = require("@solana/web3.js");
let _solana_mobile_mobile_wallet_adapter_protocol = require("@solana-mobile/mobile-wallet-adapter-protocol");
let bs58 = require("bs58");
bs58 = __toESM(bs58);
let js_base64 = require("js-base64");
//#region src/transact.ts
function getPayloadFromTransaction(transaction) {
	return (0, js_base64.fromUint8Array)("version" in transaction ? transaction.serialize() : transaction.serialize({
		requireAllSignatures: false,
		verifySignatures: false
	}));
}
function getTransactionFromWireMessage(byteArray) {
	const messageOffset = byteArray[0] * _solana_web3_js.SIGNATURE_LENGTH_IN_BYTES + 1;
	if (_solana_web3_js.VersionedMessage.deserializeMessageVersion(byteArray.slice(messageOffset, byteArray.length)) === "legacy") return _solana_web3_js.Transaction.from(byteArray);
	else return _solana_web3_js.VersionedTransaction.deserialize(byteArray);
}
async function transact(callback, config) {
	const augmentedCallback = (wallet) => {
		return callback(augmentWalletAPI(wallet));
	};
	return await (0, _solana_mobile_mobile_wallet_adapter_protocol.transact)(augmentedCallback, config);
}
async function startRemoteScenario(config) {
	const { wallet, close, associationUrl } = await (0, _solana_mobile_mobile_wallet_adapter_protocol.startRemoteScenario)(config);
	return {
		wallet: wallet.then((wallet) => {
			return augmentWalletAPI(wallet);
		}),
		close,
		associationUrl
	};
}
function augmentWalletAPI(wallet) {
	return new Proxy({}, {
		get(target, p) {
			if (target[p] == null) switch (p) {
				case "signAndSendTransactions":
					target[p] = async function({ minContextSlot, commitment, skipPreflight, maxRetries, waitForCommitmentToSendNextTransaction, transactions, ...rest }) {
						const payloads = transactions.map(getPayloadFromTransaction);
						const options = {
							min_context_slot: minContextSlot,
							commitment,
							skip_preflight: skipPreflight,
							max_retries: maxRetries,
							wait_for_commitment_to_send_next_transaction: waitForCommitmentToSendNextTransaction
						};
						const { signatures: base64EncodedSignatures } = await wallet.signAndSendTransactions({
							...rest,
							...Object.values(options).some((element) => element != null) ? { options } : null,
							payloads
						});
						return base64EncodedSignatures.map(js_base64.toUint8Array).map(bs58.default.encode);
					};
					break;
				case "signMessages":
					target[p] = async function({ payloads, ...rest }) {
						const base64EncodedPayloads = payloads.map(js_base64.fromUint8Array);
						const { signed_payloads: base64EncodedSignedMessages } = await wallet.signMessages({
							...rest,
							payloads: base64EncodedPayloads
						});
						return base64EncodedSignedMessages.map(js_base64.toUint8Array);
					};
					break;
				case "signTransactions":
					target[p] = async function({ transactions, ...rest }) {
						const payloads = transactions.map(getPayloadFromTransaction);
						const { signed_payloads: base64EncodedCompiledTransactions } = await wallet.signTransactions({
							...rest,
							payloads
						});
						return base64EncodedCompiledTransactions.map(js_base64.toUint8Array).map(getTransactionFromWireMessage);
					};
					break;
				default:
					target[p] = wallet[p];
					break;
			}
			return target[p];
		},
		defineProperty() {
			return false;
		},
		deleteProperty() {
			return false;
		}
	});
}
//#endregion
exports.startRemoteScenario = startRemoteScenario;
exports.transact = transact;
