// @trezor/protocol/lib/protocol-thp/messages/protobufDefinitions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getProtobufDefinitions = void 0;
const getProtobufDefinitions = () => ({
  ThpPairingMethod: {
    edition: 'proto2',
    values: {
      SkipPairing: 1,
      CodeEntry: 2,
      QrCode: 3,
      NFC: 4
    }
  },
  ThpDeviceProperties: {
    edition: 'proto2',
    fields: {
      internal_model: {
        rule: 'required',
        type: 'string',
        id: 1
      },
      model_variant: {
        type: 'uint32',
        id: 2,
        options: {
          default: 0
        }
      },
      protocol_version_major: {
        rule: 'required',
        type: 'uint32',
        id: 3
      },
      protocol_version_minor: {
        rule: 'required',
        type: 'uint32',
        id: 4
      },
      pairing_methods: {
        rule: 'repeated',
        type: 'ThpPairingMethod',
        id: 5
      }
    }
  },
  ThpHandshakeCompletionReqNoisePayload: {
    edition: 'proto2',
    fields: {
      host_pairing_credential: {
        type: 'bytes',
        id: 1
      }
    }
  },
  ThpCreateNewSession: {
    edition: 'proto2',
    fields: {
      passphrase: {
        type: 'string',
        id: 1
      },
      on_device: {
        type: 'bool',
        id: 2,
        options: {
          default: false
        }
      },
      derive_cardano: {
        type: 'bool',
        id: 3,
        options: {
          default: false
        }
      }
    }
  },
  ThpPairingRequest: {
    edition: 'proto2',
    fields: {
      host_name: {
        rule: 'required',
        type: 'string',
        id: 1
      },
      app_name: {
        rule: 'required',
        type: 'string',
        id: 2
      }
    }
  },
  ThpPairingRequestApproved: {
    edition: 'proto2',
    fields: {}
  },
  ThpSelectMethod: {
    edition: 'proto2',
    fields: {
      selected_pairing_method: {
        rule: 'required',
        type: 'ThpPairingMethod',
        id: 1
      }
    }
  },
  ThpPairingPreparationsFinished: {
    edition: 'proto2',
    fields: {}
  },
  ThpCodeEntryCommitment: {
    edition: 'proto2',
    fields: {
      commitment: {
        rule: 'required',
        type: 'bytes',
        id: 1
      }
    }
  },
  ThpCodeEntryChallenge: {
    edition: 'proto2',
    fields: {
      challenge: {
        rule: 'required',
        type: 'bytes',
        id: 1
      }
    }
  },
  ThpCodeEntryCpaceTrezor: {
    edition: 'proto2',
    fields: {
      cpace_trezor_public_key: {
        rule: 'required',
        type: 'bytes',
        id: 1
      }
    }
  },
  ThpCodeEntryCpaceHostTag: {
    edition: 'proto2',
    fields: {
      cpace_host_public_key: {
        rule: 'required',
        type: 'bytes',
        id: 1
      },
      tag: {
        rule: 'required',
        type: 'bytes',
        id: 2
      }
    }
  },
  ThpCodeEntrySecret: {
    edition: 'proto2',
    fields: {
      secret: {
        rule: 'required',
        type: 'bytes',
        id: 1
      }
    }
  },
  ThpQrCodeTag: {
    edition: 'proto2',
    fields: {
      tag: {
        rule: 'required',
        type: 'bytes',
        id: 1
      }
    }
  },
  ThpQrCodeSecret: {
    edition: 'proto2',
    fields: {
      secret: {
        rule: 'required',
        type: 'bytes',
        id: 1
      }
    }
  },
  ThpNfcTagHost: {
    edition: 'proto2',
    fields: {
      tag: {
        rule: 'required',
        type: 'bytes',
        id: 1
      }
    }
  },
  ThpNfcTagTrezor: {
    edition: 'proto2',
    fields: {
      tag: {
        rule: 'required',
        type: 'bytes',
        id: 1
      }
    }
  },
  ThpCredentialRequest: {
    edition: 'proto2',
    fields: {
      host_static_public_key: {
        rule: 'required',
        type: 'bytes',
        id: 1
      },
      autoconnect: {
        type: 'bool',
        id: 2,
        options: {
          default: false
        }
      },
      credential: {
        type: 'bytes',
        id: 3
      }
    }
  },
  ThpCredentialResponse: {
    edition: 'proto2',
    fields: {
      trezor_static_public_key: {
        rule: 'required',
        type: 'bytes',
        id: 1
      },
      credential: {
        rule: 'required',
        type: 'bytes',
        id: 2
      }
    }
  },
  ThpEndRequest: {
    edition: 'proto2',
    fields: {}
  },
  ThpEndResponse: {
    edition: 'proto2',
    fields: {}
  },
  MessageType: {
    edition: 'proto2',
    options: {
      '(has_bitcoin_only_values)': true,
      '(wire_enum)': true
    },
    values: {
      ThpCreateNewSession: 1000,
      ThpCredentialRequest: 1016,
      ThpCredentialResponse: 1017,
      ThpPairingRequest: 1008,
      ThpPairingRequestApproved: 1009,
      ThpSelectMethod: 1010,
      ThpPairingPreparationsFinished: 1011,
      ThpEndRequest: 1018,
      ThpEndResponse: 1019,
      ThpCodeEntryCommitment: 1024,
      ThpCodeEntryChallenge: 1025,
      ThpCodeEntryCpaceTrezor: 1026,
      ThpCodeEntryCpaceHostTag: 1027,
      ThpCodeEntrySecret: 1028,
      ThpQrCodeTag: 1032,
      ThpQrCodeSecret: 1033,
      ThpNfcTagHost: 1040,
      ThpNfcTagTrezor: 1041
    }
  }
});
exports.getProtobufDefinitions = getProtobufDefinitions;
