// @solana-mobile/mobile-wallet-adapter-protocol-web3js/lib/esm/index.browser.js
import { SIGNATURE_LENGTH_IN_BYTES, Transaction, VersionedMessage, VersionedTransaction } from "@solana/web3.js";
import { startRemoteScenario as startRemoteScenario$1, transact as transact$1 } from "@solana-mobile/mobile-wallet-adapter-protocol";
import bs58 from "bs58";
//#region src/base64Utils.ts
function fromUint8Array(byteArray) {
	return window.btoa(String.fromCharCode.call(null, ...byteArray));
}
function toUint8Array(base64EncodedByteArray) {
	return new Uint8Array(window.atob(base64EncodedByteArray).split("").map((c) => c.charCodeAt(0)));
}
//#endregion
//#region src/transact.ts
function getPayloadFromTransaction(transaction) {
	return fromUint8Array("version" in transaction ? transaction.serialize() : transaction.serialize({
		requireAllSignatures: false,
		verifySignatures: false
	}));
}
function getTransactionFromWireMessage(byteArray) {
	const messageOffset = byteArray[0] * SIGNATURE_LENGTH_IN_BYTES + 1;
	if (VersionedMessage.deserializeMessageVersion(byteArray.slice(messageOffset, byteArray.length)) === "legacy") return Transaction.from(byteArray);
	else return VersionedTransaction.deserialize(byteArray);
}
async function transact(callback, config) {
	const augmentedCallback = (wallet) => {
		return callback(augmentWalletAPI(wallet));
	};
	return await transact$1(augmentedCallback, config);
}
async function startRemoteScenario(config) {
	const { wallet, close, associationUrl } = await startRemoteScenario$1(config);
	return {
		wallet: wallet.then((wallet) => {
			return augmentWalletAPI(wallet);
		}),
		close,
		associationUrl
	};
}
function augmentWalletAPI(wallet) {
	return new Proxy({}, {
		get(target, p) {
			if (target[p] == null) switch (p) {
				case "signAndSendTransactions":
					target[p] = async function({ minContextSlot, commitment, skipPreflight, maxRetries, waitForCommitmentToSendNextTransaction, transactions, ...rest }) {
						const payloads = transactions.map(getPayloadFromTransaction);
						const options = {
							min_context_slot: minContextSlot,
							commitment,
							skip_preflight: skipPreflight,
							max_retries: maxRetries,
							wait_for_commitment_to_send_next_transaction: waitForCommitmentToSendNextTransaction
						};
						const { signatures: base64EncodedSignatures } = await wallet.signAndSendTransactions({
							...rest,
							...Object.values(options).some((element) => element != null) ? { options } : null,
							payloads
						});
						return base64EncodedSignatures.map(toUint8Array).map(bs58.encode);
					};
					break;
				case "signMessages":
					target[p] = async function({ payloads, ...rest }) {
						const base64EncodedPayloads = payloads.map(fromUint8Array);
						const { signed_payloads: base64EncodedSignedMessages } = await wallet.signMessages({
							...rest,
							payloads: base64EncodedPayloads
						});
						return base64EncodedSignedMessages.map(toUint8Array);
					};
					break;
				case "signTransactions":
					target[p] = async function({ transactions, ...rest }) {
						const payloads = transactions.map(getPayloadFromTransaction);
						const { signed_payloads: base64EncodedCompiledTransactions } = await wallet.signTransactions({
							...rest,
							payloads
						});
						return base64EncodedCompiledTransactions.map(toUint8Array).map(getTransactionFromWireMessage);
					};
					break;
				default:
					target[p] = wallet[p];
					break;
			}
			return target[p];
		},
		defineProperty() {
			return false;
		},
		deleteProperty() {
			return false;
		}
	});
}
//#endregion
export { startRemoteScenario, transact };
